File: src/main/java/org/example/bankramenserver/global/jwt/JwtAuthenticationFilter.java
package org.example.bankramenserver.global.jwt;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.example.bankramenserver.domain.auth.service.JwtService;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;
import java.util.List;
import java.util.UUID;

@Component
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {

    private final JwtService jwtService;

    private static final String[] PERMIT_URLS = {
            "/auth/kakao/login",
            "/auth/kakao/callback",
            "/swagger-ui",
            "/v3/api-docs",
            "/swagger-resources",
            "/webjars"
    };

    @Override
    protected boolean shouldNotFilter(HttpServletRequest request) {
        String path = request.getRequestURI();

        if ("/mcp".equals(path)) {
            return true;
        }

        for (String permit : PERMIT_URLS) {
            if (path.startsWith(permit)) {
                return true;
            }
        }
        return false;
    }

    private String resolveToken(HttpServletRequest request) {
        String bearer = request.getHeader("Authorization");

        if (StringUtils.hasText(bearer) && bearer.startsWith("Bearer ")) {
            return bearer.substring(7);
        }
        return null;
    }

    @Override
    protected void doFilterInternal(
            HttpServletRequest request,
            HttpServletResponse response,
            FilterChain filterChain
    ) throws ServletException, IOException {

        String token = resolveToken(request);

        if (token != null) {
            UUID userId = jwtService.validateAccessToken(token);

            UsernamePasswordAuthenticationToken authentication =
                    new UsernamePasswordAuthenticationToken(userId, null, List.of());

            SecurityContextHolder.getContext().setAuthentication(authentication);
        }

        filterChain.doFilter(request, response);
    }
}

File: src/main/kotlin/org/example/bankramenserver/global/jwt/JwtAuthenticationFilter.kt
package org.example.bankramenserver.global.jwt

import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.example.bankramenserver.domain.auth.service.JwtService
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils
import org.springframework.web.filter.OncePerRequestFilter
import java.util.UUID

@Component
class JwtAuthenticationFilter(private val jwtService: JwtService) : OncePerRequestFilter() {
    override fun shouldNotFilter(request: HttpServletRequest): Boolean { val path = request.requestURI; return path == "/mcp" || PERMIT_URLS.any(path::startsWith) }
    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) { resolveToken(request)?.let { token -> val userId: UUID = jwtService.validateAccessToken(token); SecurityContextHolder.getContext().authentication = UsernamePasswordAuthenticationToken(userId, null, emptyList()) }; filterChain.doFilter(request, response) }
    private fun resolveToken(request: HttpServletRequest): String? = request.getHeader("Authorization")?.takeIf { StringUtils.hasText(it) && it.startsWith("Bearer ") }?.substring(7)
    companion object { private val PERMIT_URLS = arrayOf("/auth/kakao/login", "/auth/kakao/callback", "/swagger-ui", "/v3/api-docs", "/swagger-resources", "/webjars") }
}
